### PR TITLE
ML TF-IDF

### PR DESCRIPTION
**Pull request by GoEddie**
Classes needed to complete the TF-IDF example: https://spark.apache.org/docs/latest/ml-features#tf-idf that completes part of #381 

- Adds Load/Save to Bucketizer
- HashingTF
- IDF
- IDFModel
- Tokenizer

### CODE REVIEW

**Comment by imback82** [comment]
@GoEddie Sorry for the delay reviewing this. I will get to this this week. Thanks!

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/TokenizerTests.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        }
+
+        [Fact]
+        public void TestTokenizer()
+        {
+            var expectedUid = "theUid";
+            var expectedInputCol = "input_col";
+            var expectedOutputCol = "output_col";
+            
+            DataFrame input = _spark.Sql("SELECT 'hello I AM a string TO, TOKENIZE' as input_col" +
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/TokenizerTests.cs]:
```suggestion
            DataFrame input = _spark.Sql("SELECT 'hello I AM a string TO, TOKENIZE' as input_col" +
                " from range(100)");
```

Next line could be removed.

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Sql;
+using Microsoft.Spark.Sql.Types;
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: Can you please remove the unnecessary namespaces?

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    /// A <see cref="HashingTF"/> Maps a sequence of terms to their term frequencies using the
+    /// hashing trick. Currently we use Austin Appleby's MurmurHash 3 algorithm
+    /// (MurmurHash3_x86_32) to calculate the hash code value for the term object. Since a simple
+    /// modulo is used to transform the hash function to a column index, it is advisable to use a
+    /// power of two as the numFeatures parameter; otherwise the features will not be mapped evenly
+    /// to the columns.
+    /// </summary>
+    public class HashingTF : IJvmObjectReferenceProvider
+    {
+
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: Please remove the extra line.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 4]
```diff
@@ -0,0 +1,166 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: Can you please remove the unnecessary namespaces?

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+using Microsoft.Spark.Interop;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Sql;
+
+namespace Microsoft.Spark.ML.Feature
+{
+    /// <summary>
+    /// Inverse document frequency (IDF). The standard formulation is used:
+    ///     idf = log((m + 1) / (d(t) + 1)), where m is the total number of documents and d(t) is
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]:
```suggestion
    /// idf = log((m + 1) / (d(t) + 1)), where m is the total number of documents and d(t) is
```

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: Can you please remove the unnecessary namespaces?

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: Can you please remove the unnecessary namespaces?

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace Microsoft.Spark.ML.Feature
+{
+    /// <summary>
+    /// A <see cref="Tokenizer"/> that converts the input string to lowercase and then splits it by
+    /// white spaces.
+    /// </summary>
+    public class Tokenizer : IJvmObjectReferenceProvider
+    {
+
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: Please remove the extra line.

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs [hunk 9]
```diff
@@ -49,12 +50,20 @@ public void TestBucketizer()
...
+                var loadedBucketizer = Bucketizer.Load(savePath);
+                Assert.Equal(bucketizer.Uid(), loadedBucketizer.Uid());
+            }
         }
 
         [Fact]
         public void TestBucketizer_MultipleColumns()
         {
-            double[][] expectedSplitsArray = new[]
+            var expectedSplitsArray = new[]
```
[the remark this answers is not in the document]
Comment (GoEddie) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: When I wrote the original bucketizer tests the feedback for the PR was to use var in tests rather than the explicit type, I don't mind which it is.

[the remark this answers is not in the document]
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: Okay I see. Cause I saw the other tests (not ML) have explicit type. But I think that's fine. Sorry for the confusion.

[the remark this answers is not in the document]
Comment (GoEddie) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: I had thought that explicit types were preferred everywhere

[the remark this answers is not in the document]
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: @imback82, should we use explicit types or `var` is fine?

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/HashingTFTests.cs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+                .SetInputCol(expectedInputCol)
+                .SetOutputCol(expectedOutputCol);
+
+            Assert.Equal(expectedFeatures, hashingTf.GetNumFeatures());
+            Assert.Equal(expectedInputCol, hashingTf.GetInputCol());
+            Assert.Equal(expectedOutputCol, hashingTf.GetOutputCol());
+
+            DataFrame input = _spark.Sql("SELECT array('this', 'is', 'a', 'string', 'a', 'a')" +
+                " as input_col");
+                                            " as input_col");
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/HashingTFTests.cs]: Need to be removed.

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/TokenizerTests.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                " from range(100)");
+                                                " from range(100)");
```
Comment (elvaliuliuliu) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/TokenizerTests.cs]: Need to be removed.

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (imback82) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: @GoEddie / @elvaliuliuliu I updated the coding guideline regarding the use of `var`: https://github.com/dotnet/spark/pull/417. I think there should be no confusion now. Thanks!

**Comment by imback82** [comment]
@GoEddie Please ping me back when you remove WIP from the title. (I don't get a notification when titles are changed). Thanks!

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/IDFModelTests.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            DataFrame featurizedData = hashingTF.Transform(wordsData);
+    
+            var idf = new IDF()
+                .SetInputCol(expectedInputCol)
+                .SetOutputCol(expectedOutputCol)
+                .SetMinDocFreq(expectedDocFrequency);
+            
+            var idfModel = idf.Fit(featurizedData);
+
+            DataFrame rescaledData = idfModel.Transform(featurizedData);
```
[the remark this answers is not in the document]
Comment (GoEddie) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/IDFModelTests.cs]: added check

File: src/csharp/Microsoft.Spark/Sql/Types/DataType.cs [hunk 13]
```diff
@@ -160,6 +160,14 @@ internal static DataType ParseDataType(JToken json)
                     }
                     else if (typeName == "udt")
                     {
+                        if (typeJObject.TryGetValue("class", out JToken classToken))
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Which function is triggering this code path?

Comment (GoEddie) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: The DataFrame that held the output from `HashingTF.Transform` caused an exception any time that `Columns()` or `Schema()` were called:

https://github.com/dotnet/spark/blob/f52f6eb832674d5b91ec87d988bafeb7f20f9f7c/src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/HashingTFTests.cs#L47

The JObject does have a class field so we can limit it to specific classes (in this case VectorUDF) but I thought that if the type had a `sqlType` then it was a good bet that we could use that.

The json include class which is something like "pyspark.ml.linalg.VectorUDT".

Comment (GoEddie) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Hi @imback82 is this ok or do you think I should do something different?

Comment (imback82) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Sorry for getting back to you late. I just took a look at this and the approach looks good to me.

FYI, the following will not work `outputVector.Collect()` because `ArrayType` is not implemented yet. Adding @elvaliuliuliu who can help implementing this.

Also for future reference, I am attaching JSON for this type:
```JSON
{
  "type": "struct",
  "fields": [
    {
      "name": "output_col",
      "type": {
        "type": "udt",
        "class": "org.apache.spark.ml.linalg.VectorUDT",
        "pyClass": "pyspark.ml.linalg.VectorUDT",
        "sqlType": {
          "type": "struct",
          "fields": [
            {
              "name": "type",
              "type": "byte",
              "nullable": false,
              "metadata": {}
            },
            {
              "name": "size",
              "type": "integer",
              "nullable": true,
              "metadata": {}
            },
            {
              "name": "indices",
              "type": {
                "type": "array",
                "elementType": "integer",
                "containsNull": false
              },
              "nullable": true,
              "metadata": {}
            },
            {
              "name": "values",
              "type": {
                "type": "array",
                "elementType": "double",
                "containsNull": false
              },
              "nullable": true,
              "metadata": {}
            }
          ]
        }
      },
      "nullable": true,
      "metadata": {
        "ml_attr": {
          "num_attrs": 10
        }
      }
    }
  ]
}
```

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs [hunk 14]
```diff
@@ -47,12 +49,20 @@ public void TestBucketizer()
             Assert.Equal(expectedInputCol, bucketizer.GetInputCol());
             Assert.Equal(expectedOutputCol, bucketizer.GetOutputCol());
             Assert.Equal(expectedSplits, bucketizer.GetSplits());
+            
+            using (var tempDirectory = new TemporaryDirectory())
+            {
+                var savePath = Path.Join(tempDirectory.Path, "bucket");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/ML/Feature/BucketizerTests.cs]: Can you go through `var` usages (https://github.com/dotnet/spark/blob/master/docs/coding-guidelines/csharp-coding-style.md)? Thanks!

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        /// <returns><see cref="HashingTF"/></returns>
+        public HashingTF SetOutputCol(string value)
+        {
+            return WrapAsHashingTF(_jvmObject.Invoke("setOutputCol", value));
+        }
+
+        /// <summary>
+        /// Gets the number of features that should be used
+        /// </summary>
+        /// <returns>int</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: Can we have a bit more descriptive comment?

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(_javaClassName, uid);
+        }
+        
+        internal HashingTF(JvmObjectReference jvmObject)
+        {
+            _jvmObject = jvmObject;
+        }
+
+        private readonly JvmObjectReference _jvmObject;
+        private const string _javaClassName = "org.apache.spark.ml.feature.HashingTF";
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: `JavaClassName`. Can you also fix this in `HashingTF`?

Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: `JavaClassName`

Comment (imback82) [reply src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: Actually, to be consistnet, can this be `private static readonly string s_hashingTFClassName = "blah"`
You can search the codebase with `ClassName =` to see examples.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        {
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(JavaClassName, uid);
+        }
+        
+        internal IDF(JvmObjectReference jvmObject)
+        {
+            _jvmObject = jvmObject;
+        }
+
+        private const string JavaClassName = "org.apache.spark.ml.feature.IDF";
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: let's define this at the top to be consistent (you can check other classes).

File: src/csharp/Microsoft.Spark/Sql/Types/DataType.cs [hunk 18]
```diff
@@ -160,6 +160,14 @@ internal static DataType ParseDataType(JToken json)
...
+                        {
+                            if (typeJObject.TryGetValue("sqlType", out JToken sqlTypeToken))
+                            {
+                                return new StructType(sqlTypeToken as JObject);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: `return new StructType((JObject)sqlTypeToken);` so that we get a better error message.

File: src/csharp/Microsoft.Spark/Sql/Types/DataType.cs (hunk 13 above)
Comment (GoEddie) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Thanks @imback82 - do I need to do something with the json?

File: src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs [hunk 19]
```diff
@@ -71,7 +73,16 @@ private DataType FromJson(JObject json)
 
         internal override bool NeedConversion() => true;
 
-        internal override object FromInternal(object obj) => throw new NotImplementedException();
+        internal override object FromInternal(object obj)
+        {
+            switch (obj)
+            {
+                case ArrayList objArrayList:
```
Comment (GoEddie) [inline src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs]: @elvaliuliuliu - i've put a simple fix in to be able to run collect() on a DataFrame that has a row with an ArrayList - if there is a better way to do this or if something else is more appropriate please let me know

Comment (imback82) [reply src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs]: @GoEddie To keep the PR small (it's already big :)), can we fix this in a separate PR? I think @elvaliuliuliu may have already started on this work.

File: src/csharp/Microsoft.Spark/Sql/Types/DataType.cs (hunk 13 above)
Comment (imback82) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Nope. I just put it as FYI.

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+        /// <returns>string UID identifying the <see cref="IDFModel"/></returns>
+        public string Uid()
+        {
+            return (string)_jvmObject.Invoke("uid");
+        }
+        
+        /// <summary>
+        /// Loads the <see cref="IDFModel"/> that was previously saved using Save
+        /// </summary>
+        /// <param name="path"></param>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: Can you fill in missing comments please? (I see them in other places as well)

File: src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs (hunk 19 above)
Comment (elvaliuliuliu) [reply src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs]: Yes, I agree with @imback82. Let's handle this threw exception for now and add a comment. I will update it once the `ArrayType` is supported.

Comment (GoEddie) [reply src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs]: Sure will take it back out.

File: src/csharp/Microsoft.Spark/Sql/Types/DataType.cs (hunk 13 above)
Comment (GoEddie) [reply src/csharp/Microsoft.Spark/Sql/Types/DataType.cs]: Great (phew)!

**Comment by GoEddie** [comment]
@imback82 I don't think there is anything left for me on this to do, if there is let me know (just messaging because I am not sure if you are waiting on me?)

**Comment by imback82** [comment]
@GoEddie Sorry for the delay. I will get to this this week. Thanks!

File: src/csharp/Microsoft.Spark/ML/Feature/Bucketizer.cs [hunk 21]
```diff
@@ -243,5 +254,17 @@ public Bucketizer SetHandleInvalid(string value)
...
             return WrapAsBucketizer(_jvmObject.Invoke("setHandleInvalid", value.ToString()));
         }
+        
+        /// <summary>
+        /// The reference we get back from each call isn't usable unless we wrap it in a new dotnet
+        /// <see cref="Bucketizer"/>
+        /// </summary>
+        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
+        /// <see cref="Bucketizer"/></param>
+        /// <returns><see cref="Bucketizer"/></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/Bucketizer.cs]: missing comment.

Comment (imback82) [reply src/csharp/Microsoft.Spark/ML/Feature/Bucketizer.cs]: or let's remove this comment block since it's a private function:
```C#
private static Bucketizer WrapAsBucketizer(object obj) => new Bucketizer((JvmObjectReference)obj);
```
(Check `DataFrameNaFunctions.cs` for example)

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 22]
```diff
@@ -0,0 +1,192 @@
...
+        }
+
+        /// <summary>
+        /// The reference we get back from each call isn't usable unless we wrap it in a new dotnet
+        /// <see cref="HashingTF"/>
+        /// </summary>
+        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
+        /// <see cref="HashingTF"/></param>
+        /// <returns><see cref="HashingTF"/></returns>
+        private static HashingTF WrapAsHashingTF(object obj)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: ditto.

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 23]
```diff
@@ -0,0 +1,192 @@
...
+
+        /// <summary>
+        /// Loads the <see cref="HashingTF"/> that was previously saved using Save
+        /// </summary>
+        /// <param name="path">The path the previous <see cref="HashingTF"/> was saved to</param>
+        /// <returns><see cref="HashingTF"/></returns>
+        public static HashingTF Load(string path)
+        {
+            return WrapAsHashingTF(
+                SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_hashingTfClassName,"load", path));
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: space after `,`

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+            return WrapAsHashingTF(_jvmObject.Invoke("setOutputCol", value));
+        }
+
+        /// <summary>
+        /// Gets the number of features that should be used. Since a simple modulo is used to
+        /// transform the hash function to a column index, it is advisable to use a power of two
+        /// as the numFeatures parameter; otherwise the features will not be mapped evenly to the
+        /// columns.
+        /// </summary>
+        /// <returns>int</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: just `int`? Can this be a bit more descriptive?

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+        /// <returns><see cref="HashingTF"/></returns>
+        public HashingTF SetNumFeatures(int value)
+        {
+            return WrapAsHashingTF(_jvmObject.Invoke("setNumFeatures", value));
+        }
+
+        /// <summary>
+        /// An immutable unique ID for the object and its derivatives.
+        /// </summary>
+        /// <returns>string</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+        {
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(s_IDFClassName, uid);
+        }
+        
+        internal IDF(JvmObjectReference jvmObject)
+        {
+            _jvmObject = jvmObject;
+        }
+
+        private readonly JvmObjectReference _jvmObject;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: Can we make this consistent with existing code? I think we declare private variables first. Check https://github.com/dotnet/spark/blob/master/src/csharp/Microsoft.Spark/Sql/DataFrameWriter.cs for example.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 27]
```diff
@@ -0,0 +1,163 @@
...
+        /// <returns><see cref="IDF"/></returns>
+        public IDF SetOutputCol(string value)
+        {
+            return WrapAsIDF(_jvmObject.Invoke("setOutputCol", value));
+        }
+
+        /// <summary>
+        /// Minimum of documents in which a term should appear for filtering
+        /// </summary>
+        /// <returns>int</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: ditto.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 28]
```diff
@@ -0,0 +1,163 @@
...
+        public int GetMinDocFreq()
+        {
+            return (int)_jvmObject.Invoke("getMinDocFreq");
+        }
+        
+        /// <summary>
+        /// Minimum of documents in which a term should appear for filtering
+        /// </summary>
+        /// <param name="value">int, the minimum of documents a term should appear in</param>
+        /// <returns></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: missing comment.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+        public IDF SetMinDocFreq(int value)
+        {
+            return WrapAsIDF(_jvmObject.Invoke("setMinDocFreq", value));
+        }
+        
+        /// <summary>
+        /// Fits a model to the input data.
+        /// </summary>
+        /// <param name="source">The <see cref="DataFrame"/> to fit the model to</param>
+        /// <returns><see cref="IDFModel"/></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: maybe new IDFModel object? same in other places.

File: src/csharp/Microsoft.Spark/ML/Feature/IDF.cs [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+            return WrapAsIDF(_jvmObject.Invoke("save", path));
+        }
+
+        /// <summary>
+        /// The reference we get back from each call isn't usable unless we wrap it in a new dotnet
+        /// <see cref="IDF"/>
+        /// </summary>
+        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
+        /// <see cref="IDF"/></param>
+        /// <returns><see cref="IDF"/></returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDF.cs]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+        {
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(s_IDFModelClassName, uid);
+        }
+        
+        internal IDFModel(JvmObjectReference jvmObject)
+        {
+            _jvmObject = jvmObject;
+        }
+
+        private readonly JvmObjectReference _jvmObject;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+        }
+        
+        /// <summary>
+        /// The reference we get back from each call isn't usable unless we wrap it in a new dotnet
+        /// <see cref="IDFModel"/>
+        /// </summary>
+        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
+        /// <see cref="IDFModel"/></param>
+        /// <returns><see cref="IDFModel"/></returns>
+        private static IDFModel WrapAsIDFModel(object obj)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 33]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            _jvmObject = SparkEnvironment.JvmBridge.CallConstructor(s_tokenizerClassName, uid);
+        }
+        
+        internal Tokenizer(JvmObjectReference jvmObject)
+        {
+            _jvmObject = jvmObject;
+        }
+
+        private readonly JvmObjectReference _jvmObject;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: ditto

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 34]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        
+        /// <summary>
+        /// The reference we get back from each call isn't usable unless we wrap it in a new dotnet
+        /// <see cref="Tokenizer"/>
+        /// </summary>
+        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
+        /// <see cref="Tokenizer"/></param>
+        /// <returns><see cref="Tokenizer"/></returns>
+        private static Tokenizer WrapAsTokenizer(object obj)
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: ditto

File: src/csharp/Microsoft.Spark/Microsoft.Spark.csproj [hunk 35]
```diff
@@ -33,10 +33,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <Content Include="..\..\scala\microsoft-spark-*\target\microsoft-spark-*.jar"
-             Link="jars\%(Filename)%(Extension)"
-             Pack="true"
-             PackagePath="jars\%(Filename)%(Extension)" />
+    <Content Include="..\..\scala\microsoft-spark-*\target\microsoft-spark-*.jar" Link="jars\%(Filename)%(Extension)" Pack="true" PackagePath="jars\%(Filename)%(Extension)" />
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Microsoft.Spark.csproj]: can we revert this change?

File: src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs [hunk 36]
```diff
@@ -3,7 +3,9 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Collections;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/Sql/Types/ComplexTypes.cs]: revert these changes?

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 37]
```diff
@@ -138,7 +139,7 @@ public HashingTF SetOutputCol(string value)
         /// as the numFeatures parameter; otherwise the features will not be mapped evenly to the
         /// columns.
         /// </summary>
-        /// <returns>int</returns>
+        /// <returns>int, the number of features to be used</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: `The number of features to be used` should be enough.

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 38]
```diff
@@ -176,17 +177,8 @@ public DataFrame Transform(DataFrame source)
...
-        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
-        /// <see cref="HashingTF"/></param>
-        /// <returns><see cref="HashingTF"/></returns>
-        private static HashingTF WrapAsHashingTF(object obj)
-        {
-            return new HashingTF((JvmObjectReference)obj);
-        }
+        
+        private static HashingTF WrapAsHashingTF(object obj) 
+            => new HashingTF((JvmObjectReference)obj);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: Let's put the `=>` in the previous line similar to `Catalog.cs`.
```C#
private static HashingTF WrapAsHashingTF(object obj) =>
    new HashingTF((JvmObjectReference)obj);
```

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 39]
```diff
@@ -88,7 +89,7 @@ public IDFModel SetOutputCol(string value)
         /// <summary>
         /// Minimum of documents in which a term should appear for filtering
         /// </summary>
-        /// <returns>int</returns>
+        /// <returns>int, minimum number of documents a term should appear</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: Ditto

File: src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs [hunk 40]
```diff
@@ -132,22 +133,13 @@ public static IDFModel Load(string path)
...
-        /// </summary>
-        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
-        /// <see cref="IDFModel"/></param>
-        /// <returns><see cref="IDFModel"/></returns>
-        private static IDFModel WrapAsIDFModel(object obj)
-        {
-            return new IDFModel((JvmObjectReference)obj);
-        }
+        private static IDFModel WrapAsIDFModel(object obj) 
+            => new IDFModel((JvmObjectReference)obj);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/IDFModel.cs]: Ditto

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 41]
```diff
@@ -88,7 +89,8 @@ public Tokenizer SetOutputCol(string value)
         /// column
         /// </summary>
         /// <param name="source">The DataFrame to transform</param>
-        /// <returns><see cref="DataFrame"/></returns>
+        /// <returns>New <see cref="DataFrame"/> object with the source <see cref="DataFrame"/>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: For tags spanning multiple lines, can we follow the same as `summary`?
```
<returns>
here
</returns>
```

File: src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs [hunk 42]
```diff
@@ -121,22 +123,13 @@ public static Tokenizer Load(string path)
...
-        /// </summary>
-        /// <param name="obj">The <see cref="JvmObjectReference"/> to convert into a dotnet
-        /// <see cref="Tokenizer"/></param>
-        /// <returns><see cref="Tokenizer"/></returns>
-        private static Tokenizer WrapAsTokenizer(object obj)
-        {
-            return new Tokenizer((JvmObjectReference)obj);
-        }
+        private static Tokenizer WrapAsTokenizer(object obj) 
+            => new Tokenizer((JvmObjectReference)obj);
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/Tokenizer.cs]: Ditto

File: src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs [hunk 43]
```diff
@@ -75,7 +76,7 @@ public HashingTF Save(string path)
         /// <summary>
         /// Gets the binary toggle that controls term frequency counts
         /// </summary>
-        /// <returns>bool</returns>
+        /// <returns>bool showing term frequency counts</returns>
```
Comment (imback82) [inline src/csharp/Microsoft.Spark/ML/Feature/HashingTF.cs]: it doesn't "show" right? I think it "controls" as the summary says. `Flag that controls term frequency counts`?

**Comment by GoEddie** [comment]
Awesome :)
